site/pages/api/hootsuite-callback.api.ts

```diff
@@ -0,0 +1,112 @@
+import { getParameter, putParameter } from "../../data/ssm";
+import { initiateRefreshAuth } from "../../middleware.api";
+import { hootsuiteMeSchema, hootsuiteTokenSchema } from "../../schemas/hootsuite.schema";
+import { getSession } from "../../utils/apiUtils/auth";
+import { redirectToError, redirectTo, setCookieOnResponseObject } from "../../utils/apiUtils/index";
+
+const hootsuiteCallback = async (req: NextApiRequest, res: NextApiResponse) => {
+    try {
+        const { code, state } = req.query;
+
+        const idToken = state?.toString() ? await getParameter(`/${state.toString()}/token`) : undefined;
+        const refreshToken = state?.toString() ? await getParameter(`/${state.toString()}/refresh-token`) : undefined;
+
+        const [clientId, clientSecret] = await Promise.all([
+            getParameter(`/social/hootsuite/client_id`),
+            getParameter(`/social/hootsuite/client_secret`),
+        ]);
+
+        if (!clientId || !clientSecret) {
+            throw new Error("clientId and clientSecret must be defined");
+        }
+
+        const authKey = `${clientId.Parameter?.Value || ""}:${clientSecret.Parameter?.Value || ""}`;
+
+        const authToken = `Basic ${Buffer.from(authKey).toString("base64")}`;
+
+        if (!idToken?.Parameter?.Value) {
+            throw new Error("idToken must be defined");
+        }
+        const decodedToken = jose.decodeJwt(idToken?.Parameter?.Value);
+        const username = (decodedToken["cognito:username"] as string) ?? null;
+        if (refreshToken?.Parameter?.Value) {
+            setCookieOnResponseObject(COOKIES_REFRESH_TOKEN, refreshToken?.Parameter?.Value, res);
+
+            const refreshResult = await initiateRefreshAuth(username, refreshToken?.Parameter?.Value);
+            if (refreshResult.AuthenticationResult?.IdToken) {
+                setCookieOnResponseObject(COOKIES_ID_TOKEN, refreshResult.AuthenticationResult.IdToken, res);
+                res.setHeader(
+                    "Set-Cookie",
+                    `${COOKIES_ID_TOKEN}=${refreshResult.AuthenticationResult.IdToken}; Path=/; HttpOnly`,
+                );
+            }
+        }
+
+        const tokenResponse = await fetch(`${HOOTSUITE_URL}oauth2/token`, {
+            method: "POST",
+            body: new URLSearchParams({
+                grant_type: "authorization_code",
+                code: code?.toString() ?? "",
+                redirect_uri: `http://localhost:3000/api/hootsuite-callback`,
+            }),
+            headers: {
+                "Content-Type": "application/x-www-form-urlencoded",
+                Authorization: authToken,
+            },
+        });
+
+        if (!tokenResponse.ok) {
+            const message = `An error has occured: ${tokenResponse.status}`;
+            throw new Error(message);
+        }
+
+        const tokenResult = hootsuiteTokenSchema.parse(await tokenResponse.json());
+
+        const session = getSession(req);
+        if (!session) {
```

Comment:
Should we throw an error is the session user is not an org admin?

Reply:
done